### ISSUE TITLE
sometimes ghcid reports build is All Good when it isn't

### ISSUE THREAD

**Issue by aryairani** [ndmitchell/ghcid#207]
Then I check in code that isn't All Good. 😞 
I haven't discovered a pattern to it yet.  I am using ghcid v0.7, stack v1.7.1, and ghc 8.4.2.
Any suggestions?  Thanks in advance.

**Comment by ndmitchell** [comment]
So it normally works and sometimes doesn't? Is the type of not good consistent? No real ideas, although --verbose often helps - but not really until you know it's bust.

**Comment by aryairani** [comment]
Yes, it works wonderfully most if the time, but then even while it continues to appear to work (eg displaying errors and then reporting All Good once I've fixed them), it may be no longer be rebuilding some other files that have errors in them.  

e.g. yesterday I had started using a package that wasn't in my cabal file, and it all seemed to be compiling fine, but when I pushed to our CI server, it was obvious that it never could have been building locally.  In another case, I was missing a DoAndIfThenElse but still got no errors during my ghcid session although I would, typically.

**Comment by ndmitchell** [comment]
So `DoAndIfThenElse` is the weird kind of extension that is usually enabled by default in `ghci`, but only enabled through Cabal etc. if you specify it. I suggest you solve that by either doing `cabal repl` or similar to start your ghci session (which means DoAndIfThenElse will be disabled) or turning it on in your .cabal, possibly by changing the Haskell language to Haskell2010 or similar.

The issue of using a package not in your .cabal file would also be picked up by `cabal repl` as the ghci prompt. What is it currently? What's the first few lines of ghcid when it starts? That includes the command line it runs.

**Comment by ndmitchell** [comment]
@mitchellwrosen I believe that's an underlying GHC bug - GHC gets quite upset if you delete something it has loaded in memory. That needs patching in GHC itself.

**Comment by aryairani** [comment]
Not sure if this is related, but is there a way I can make ghcid build all the subprojects that `stack build` would build?  Currently it seems to pick one on its own (which does happen to be my one with the most code, so that's pretty good).

**Comment by ndmitchell** [comment]
What command line is ghcid guessing at? That should be the first or second line when ghcid starts up. That command line controls what ghcid loads, and if you can find a better stack command line to get different things loaded that will help. Ghcid prods a running ghci, but doesn't really deal with how that running ghci gets started.

**Comment by identicalsnowflake** [comment]
I encountered something similar. For me, the underlying cause was I was using unboxed sums, and GHCI doesn't support them (running `cabal new-repl`, even with `-fno-code`, produces a weird error where it loads half of the modules while saying it doesn't support unboxed sums), and ghcid doesn't seem to notice this, which results in it reporting `All good` while only actually checking the modules that loaded before ghci bailed out, resulting in problems to any downstream files not being noticed.

**Comment by ndmitchell** [comment]
@identicalsnowflake see #173 where I'm tracking unboxed sums/tuples separately.
